src/node_url_pattern.cc

```diff
@@ -0,0 +1,714 @@
+
+  ada::url_pattern_input input;
+  std::optional<std::string> baseURL{};
+  std::string input_base;
+  if (args.Length() == 0) {
+    input = ada::url_pattern_init{};
+  } else if (args[0]->IsString()) {
+    Utf8Value input_value(env->isolate(), args[0].As<String>());
+    input_base = input_value.ToString();
+    input = std::string_view(input_base);
+  } else if (args[0]->IsObject()) {
+    input = URLPatternInit::FromJsObject(env, args[0].As<Object>());
+  } else {
+    THROW_ERR_INVALID_ARG_TYPE(
+        env, "URLPattern input needs to be a string or an object");
+    return;
+  }
+
+  if (args.Length() > 1 && args[1]->IsString()) {
+    Utf8Value base_url_value(env->isolate(), args[1].As<String>());
+    baseURL = base_url_value.ToStringView();
+  }
+
+  std::optional<std::string_view> baseURL_opt =
+      baseURL ? std::optional<std::string_view>(*baseURL) : std::nullopt;
+  args.GetReturnValue().Set(url_pattern->Test(env, input, baseURL_opt));
+}
+
+void URLPattern::Protocol(const FunctionCallbackInfo<Value>& info) {
+  URLPattern* url_pattern;
+  ASSIGN_OR_RETURN_UNWRAP(&url_pattern, info.This());
+  Local<Value> result;
+  if (!url_pattern->Protocol().ToLocal(&result)) {
+    return;
+  }
+  info.GetReturnValue().Set(result);
+}
+
+void URLPattern::Username(const FunctionCallbackInfo<Value>& info) {
+  URLPattern* url_pattern;
+  ASSIGN_OR_RETURN_UNWRAP(&url_pattern, info.This());
+  Local<Value> result;
+  if (!url_pattern->Username().ToLocal(&result)) {
+    return;
+  }
+  info.GetReturnValue().Set(result);
+}
+
+void URLPattern::Password(const FunctionCallbackInfo<Value>& info) {
+  URLPattern* url_pattern;
+  ASSIGN_OR_RETURN_UNWRAP(&url_pattern, info.This());
+  Local<Value> result;
+  if (!url_pattern->Password().ToLocal(&result)) {
+    return;
+  }
+  info.GetReturnValue().Set(result);
+}
+
+void URLPattern::Hostname(const FunctionCallbackInfo<Value>& info) {
+  URLPattern* url_pattern;
+  ASSIGN_OR_RETURN_UNWRAP(&url_pattern, info.This());
+  Local<Value> result;
+  if (!url_pattern->Hostname().ToLocal(&result)) {
+    return;
+  }
+  info.GetReturnValue().Set(result);
+}
+
+void URLPattern::Port(const FunctionCallbackInfo<Value>& info) {
+  URLPattern* url_pattern;
+  ASSIGN_OR_RETURN_UNWRAP(&url_pattern, info.This());
+  Local<Value> result;
+  if (!url_pattern->Port().ToLocal(&result)) {
+    return;
+  }
+  info.GetReturnValue().Set(result);
+}
+
+void URLPattern::Pathname(const FunctionCallbackInfo<Value>& info) {
+  URLPattern* url_pattern;
+  ASSIGN_OR_RETURN_UNWRAP(&url_pattern, info.This());
+  Local<Value> result;
+  if (!url_pattern->Pathname().ToLocal(&result)) {
+    return;
+  }
+  info.GetReturnValue().Set(result);
+}
+
+void URLPattern::Search(const FunctionCallbackInfo<Value>& info) {
+  URLPattern* url_pattern;
+  ASSIGN_OR_RETURN_UNWRAP(&url_pattern, info.This());
+  Local<Value> result;
+  if (!url_pattern->Search().ToLocal(&result)) {
+    return;
+  }
+  info.GetReturnValue().Set(result);
+}
```

Comment:
the way these are defined, the property is going to be reserialized as a new string on every call. Should probably memoize these to avoid the extraneous/duplicative copies.

Reply:
Any suggestions on how to memoize with least amount of code and highest amount of memory safety possible?

Reply:
You would need to have `v8::Global<v8::String>` member fields to cache the values. Or, maybe use `SetLazyDataProperty()` ... not sure if the latter would make it spec compliant, tho, as I believe it makes the properties instance properties rather than prototype properties.